jersey-servers/src/main/java/com/palantir/remoting3/servers/jersey/QosExceptionMapper.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.servers.jersey;
+
+import com.google.common.net.HttpHeaders;
+import com.palantir.remoting.api.errors.QosException;
+import java.net.URL;
+import java.time.Duration;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.ExceptionMapper;
+import javax.ws.rs.ext.Provider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An {@link ExceptionMapper} that turns http-remoting {@link QosException}s into appropriate HTTP error codes and
+ * headers. Four different cases are distinguished:
+ * <ol>
+ * <li>Retry any node of this service immediately: HTTP 307 Temporary Redirect</li>
+ * <li>Retry any node of this service after a given backoff time: HTTP 429 Too Many Requests + Retry-After header</li>
+ * <li>Retry a specific (other) node of this service immediately: HTTP 308 Permanent Redirect + Location header</li>
+ * <li>Don't retry any node of this service: HTTP 503 Unavailable</li>
+ * </ol>
+ */
+@Provider
+final class QosExceptionMapper implements ExceptionMapper<QosException> {
+
+    private static final Logger log = LoggerFactory.getLogger(QosExceptionMapper.class);
+
+    @Override
+    public Response toResponse(QosException exception) {
+        log.debug("Possible quality-of-service intervention", exception);
+        if (exception instanceof QosException.Retry) {
+            QosException.Retry retry = (QosException.Retry) exception;
+            if (!retry.getBackoff().isPresent() && !retry.getRedirectTo().isPresent()) {
+                return retryAnyNow();
+            } else {
```

Comment:
left a comment on https://github.com/palantir/http-remoting-api/pull/30/files - we shouldn't allow such a Retry then